mm2src/coins/eth.rs

```diff
@@ -2420,6 +2435,128 @@ impl MmCoin for EthCoin {
+            let gas_price = increase_by_percent_one_gwei(gas_price, TRADE_PREIMAGE_GAS_PRICE_PERCENT);
+            let gas_limit = match coin.coin_type {
+                EthCoinType::Eth => {
+                    // this gas_limit includes gas for `ethPayment` and `senderRefund` contract calls
+                    U256::from(300_000)
+                },
+                EthCoinType::Erc20(_) => {
+                    let value = match value {
+                        TradePreimageValue::Exact(value) | TradePreimageValue::UpperBound(value) => {
+                            try_map!(wei_from_big_decimal(&value, coin.decimals), TradePreimageError::Other)
+                        },
+                    };
+                    let allowed = try_map!(
+                        coin.allowance(coin.swap_contract_address).compat().await,
+                        TradePreimageError::Other
+                    );
+                    if allowed < value {
+                        // this gas_limit includes gas for `approve`, `erc20Payment` and `senderRefund` contract calls
+                        U256::from(450_000)
+                    } else {
+                        // this gas_limit includes gas for `erc20Payment` and `senderRefund` contract calls
+                        U256::from(300_000)
+                    }
+                },
+            };
+
+            let total_fee = gas_limit * gas_price;
+            let amount = try_map!(u256_to_big_decimal(total_fee, 18), TradePreimageError::Other);
+            Ok(TradeFee {
+                coin: "ETH".to_owned(),
+                amount: amount.into(),
+            })
+        };
+        Box::new(fut.boxed().compat())
+    }
+
+    fn get_receiver_trade_fee(&self) -> Box<dyn Future<Item = TradeFee, Error = TradePreimageError> + Send> {
+        let coin = self.clone();
+        let fut = async move {
+            let gas_price = try_map!(coin.get_gas_price().compat().await, TradePreimageError::Other);
+            let gas_price = increase_by_percent_one_gwei(gas_price, TRADE_PREIMAGE_GAS_PRICE_PERCENT);
+            let total_fee = gas_price * U256::from(150_000);
+            let amount = try_map!(u256_to_big_decimal(total_fee, 18), TradePreimageError::Other);
+            Ok(TradeFee {
+                coin: "ETH".to_owned(),
+                amount: amount.into(),
+            })
+        };
+        Box::new(fut.boxed().compat())
+    }
+
+    fn get_fee_to_send_taker_fee(
+        &self,
+        dex_fee_amount: BigDecimal,
+    ) -> Box<dyn Future<Item = TradeFee, Error = TradePreimageError> + Send> {
+        const EXCEPTION_ERROR_CODE: i64 = -32016;
+
+        let coin = self.clone();
+        let fut = async move {
+            let dex_fee_amount = try_map!(
+                wei_from_big_decimal(&dex_fee_amount, coin.decimals),
+                TradePreimageError::Other
+            );
+
+            // pass the dummy params
+            let to_addr = Address::random();
```

Comment:
It's error-prone to use a random address here. We can occasionally get a smart contract address, which may result in RPC error on gas estimation or incorrect estimation result. I think it's worth passing the dex fee address pubkey here similarly to `SwapOps::send_taker_fee`.